File: src/main/java/no/nav/veilarbaktivitet/internapi/InternAktivitetMapper.java
package no.nav.veilarbaktivitet.internapi;

import no.nav.veilarbaktivitet.aktivitet.domain.*;
import no.nav.veilarbaktivitet.internapi.model.*;
import no.nav.veilarbaktivitet.internapi.model.Aktivitet.AktivitetBuilder;
import no.nav.veilarbaktivitet.internapi.model.Aktivitet.AktivitetTypeEnum;
import no.nav.veilarbaktivitet.stilling_fra_nav.CvKanDelesData;
import no.nav.veilarbaktivitet.stilling_fra_nav.StillingFraNavData;

import java.util.Optional;

import static no.nav.veilarbaktivitet.util.DateUtils.toLocalDate;
import static no.nav.veilarbaktivitet.util.DateUtils.toOffsetDateTime;

public class InternAktivitetMapper {

    private InternAktivitetMapper() {}

    public static Aktivitet mapTilAktivitet(AktivitetData aktivitetData) {
        AktivitetTypeData aktivitetType = aktivitetData.getAktivitetType();

        AktivitetBuilder<?, ?> builder = switch (aktivitetType) {
            case EGENAKTIVITET -> mapTilEgenaktivitet(aktivitetData);
            case JOBBSOEKING -> mapTilJobbsoeking(aktivitetData);
            case SOKEAVTALE -> mapTilSokeavtale(aktivitetData);
            case IJOBB -> mapTilIjobb(aktivitetData);
            case BEHANDLING -> mapTilBehandling(aktivitetData);
            case MOTE -> mapTilMote(aktivitetData);
            case SAMTALEREFERAT -> mapTilSamtalereferat(aktivitetData);
            case STILLING_FRA_NAV -> mapTilStillingFraNav(aktivitetData);
        };

        return builder
                .avtaltMedNav(aktivitetData.isAvtalt())
                .aktivitetId(aktivitetData.getId().toString())
                .kontorsperreEnhetId(aktivitetData.getKontorsperreEnhetId())
                .oppfolgingsperiodeId(aktivitetData.getOppfolgingsperiodeId())
                .status(Aktivitet.StatusEnum.valueOf(aktivitetData.getStatus().name()))
                .beskrivelse(aktivitetData.getBeskrivelse())
                .tittel(aktivitetData.getTittel())
                .fraDato(toOffsetDateTime(aktivitetData.getFraDato()))
                .tilDato(toOffsetDateTime(aktivitetData.getTilDato()))
                .opprettetDato(toOffsetDateTime(aktivitetData.getOpprettetDato()))
                .endretDato(toOffsetDateTime(aktivitetData.getOpprettetDato()))
                .build();
    }

    private static AktivitetBuilder<?, ?> mapTilEgenaktivitet(AktivitetData aktivitetData) {
        EgenAktivitetData egenAktivitetData = aktivitetData.getEgenAktivitetData();

        return Egenaktivitet.builder()
                .aktivitetType(AktivitetTypeEnum.EGENAKTIVITET)
                .hensikt(egenAktivitetData.getHensikt())
                .oppfolging(egenAktivitetData.getOppfolging());
    }

    private static AktivitetBuilder<?, ?> mapTilJobbsoeking(AktivitetData aktivitetData) {
        StillingsoekAktivitetData stillingsSoekAktivitetData = aktivitetData.getStillingsSoekAktivitetData();

        Optional<StillingsoekEtikettData> stillingsoekEtikett = Optional.ofNullable(stillingsSoekAktivitetData.getStillingsoekEtikett());
        return Jobbsoeking.builder()
                .aktivitetType(AktivitetTypeEnum.JOBBSOEKING)
                .arbeidsgiver(stillingsSoekAktivitetData.getArbeidsgiver())
                .stillingsTittel(stillingsSoekAktivitetData.getStillingsTittel())
                .arbeidssted(stillingsSoekAktivitetData.getArbeidssted())
                .stillingsoekEtikett(stillingsoekEtikett.map(Enum::name).map(Jobbsoeking.StillingsoekEtikettEnum::valueOf).orElse(null))
                .kontaktPerson(stillingsSoekAktivitetData.getKontaktPerson());
    }

    private static AktivitetBuilder<?, ?> mapTilSokeavtale(AktivitetData aktivitetData) {
        SokeAvtaleAktivitetData sokeAvtaleAktivitetData = aktivitetData.getSokeAvtaleAktivitetData();

        return Sokeavtale.builder()
                .aktivitetType(AktivitetTypeEnum.SOKEAVTALE)
                .antallStillingerSokes(sokeAvtaleAktivitetData.getAntallStillingerSokes())
                .antallStillingerIUken(sokeAvtaleAktivitetData.getAntallStillingerIUken())
                .avtaleOppfolging(sokeAvtaleAktivitetData.getAvtaleOppfolging());
    }

    private static AktivitetBuilder<?, ?> mapTilIjobb(AktivitetData aktivitetData) {
        IJobbAktivitetData iJobbAktivitetData = aktivitetData.getIJobbAktivitetData();

        Optional<JobbStatusTypeData> jobbStatusType = Optional.ofNullable(iJobbAktivitetData.getJobbStatusType());
        return Ijobb.builder()
                .aktivitetType(AktivitetTypeEnum.IJOBB)
                .jobbStatusType(jobbStatusType.map(Enum::name).map(Ijobb.JobbStatusTypeEnum::valueOf).orElse(null))
                .ansettelsesforhold(iJobbAktivitetData.getAnsettelsesforhold())
                .arbeidstid(iJobbAktivitetData.getArbeidstid());
    }

    private static AktivitetBuilder<?, ?> mapTilBehandling(AktivitetData aktivitetData) {
        BehandlingAktivitetData behandlingAktivitetData = aktivitetData.getBehandlingAktivitetData();

        return Behandling.builder()
                .aktivitetType(AktivitetTypeEnum.BEHANDLING)
                .behandlingType(behandlingAktivitetData.getBehandlingType())
                .behandlingSted(behandlingAktivitetData.getBehandlingSted())
                .effekt(behandlingAktivitetData.getEffekt())
                .behandlingOppfolging(behandlingAktivitetData.getBehandlingOppfolging());
    }

    private static AktivitetBuilder<?, ?> mapTilMote(AktivitetData aktivitetData) {
        MoteData moteData = aktivitetData.getMoteData();

        return Mote.builder()
                .aktivitetType(AktivitetTypeEnum.MOTE)
                .adresse(moteData.getAdresse())
                .forberedelser(moteData.getForberedelser())
                .kanal(Mote.KanalEnum.valueOf(moteData.getKanal().name()))
                .referat(moteData.getReferat())
                .referatPublisert(moteData.isReferatPublisert());
    }

    private static AktivitetBuilder<?, ?> mapTilSamtalereferat(AktivitetData aktivitetData) {
        MoteData moteData = aktivitetData.getMoteData();

        return Samtalereferat.builder()
                .aktivitetType(AktivitetTypeEnum.SAMTALEREFERAT)
                .kanal(Samtalereferat.KanalEnum.valueOf(moteData.getKanal().name()))
                .referat(moteData.getReferat())
                .referatPublisert(moteData.isReferatPublisert());
    }

    private static AktivitetBuilder<?, ?> mapTilStillingFraNav(AktivitetData aktivitetData) {
        StillingFraNavData stillingFraNavData = aktivitetData.getStillingFraNavData();
        StillingFraNavAllOfCvKanDelesData stillingFraNavCvKanDelesData = mapCvKanDelesData(stillingFraNavData.getCvKanDelesData());

        StillingFraNav.SoknadsstatusEnum soknadsstatusEnum = Optional.ofNullable(stillingFraNavData.getSoknadsstatus())
                .map(Enum::name)
                .map(StillingFraNav.SoknadsstatusEnum::fromValue)
                .orElse(null);

        return StillingFraNav.builder()
                .aktivitetType(AktivitetTypeEnum.STILLING_FRA_NAV)
                .cvKanDelesData(stillingFraNavCvKanDelesData)
                .soknadsfrist(stillingFraNavData.getSoknadsfrist())
                .svarfrist(toLocalDate(stillingFraNavData.getSvarfrist()))
                .arbeidsgiver(stillingFraNavData.getArbeidsgiver())
                .bestillingsId(stillingFraNavData.getBestillingsId())
                .stillingsId(stillingFraNavData.getStillingsId())
                .arbeidssted(stillingFraNavData.getArbeidssted())
                .soknadsstatus(soknadsstatusEnum);
    }

    private static StillingFraNavAllOfCvKanDelesData mapCvKanDelesData(CvKanDelesData cvKanDelesData) {
        if (cvKanDelesData == null) return null;

        return StillingFraNavAllOfCvKanDelesData.builder()
                .kanDeles(cvKanDelesData.getKanDeles())
                .endretTidspunkt(toOffsetDateTime(cvKanDelesData.getEndretTidspunkt()))
                .endretAv(cvKanDelesData.getEndretAv())
                .endretAvType(StillingFraNavAllOfCvKanDelesData.EndretAvTypeEnum.valueOf(cvKanDelesData.getEndretAvType().name()))
                .avtaltDato(toLocalDate(cvKanDelesData.getAvtaltDato()))
                .build();
    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/internapi/InternAktivitetMapper.kt
package no.nav.veilarbaktivitet.internapi

import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetTypeData
import no.nav.veilarbaktivitet.aktivitet.domain.IJobbAktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.JobbStatusTypeData
import no.nav.veilarbaktivitet.internapi.model.*
import no.nav.veilarbaktivitet.stilling_fra_nav.CvKanDelesData
import no.nav.veilarbaktivitet.util.DateUtils
import org.apache.commons.lang3.NotImplementedException

object InternAktivitetMapper {
    fun mapTilAktivitet(aktivitetData: AktivitetData): Aktivitet {
        val aktivitetType: AktivitetTypeData = aktivitetData.aktivitetType
        return when (aktivitetType) {
            AktivitetTypeData.EGENAKTIVITET -> mapTilEgenaktivitet(aktivitetData)
            AktivitetTypeData.JOBBSOEKING -> mapTilJobbsoeking(aktivitetData)
            AktivitetTypeData.SOKEAVTALE -> mapTilSokeavtale(aktivitetData)
            AktivitetTypeData.IJOBB -> mapTilIjobb(aktivitetData)
            AktivitetTypeData.BEHANDLING -> mapTilBehandling(aktivitetData)
            AktivitetTypeData.MOTE -> mapTilMote(aktivitetData)
            AktivitetTypeData.SAMTALEREFERAT -> mapTilSamtalereferat(aktivitetData)
            AktivitetTypeData.STILLING_FRA_NAV -> mapTilStillingFraNav(aktivitetData)
            AktivitetTypeData.EKSTERNAKTIVITET -> mapTilTiltak(aktivitetData)
        }
    }

    private fun mapTilTiltak(aktivitetData: AktivitetData): Aktivitet {
        throw NotImplementedException("TODO")
    }

    private fun mapTilEgenaktivitet(aktivitetData: AktivitetData): Egenaktivitet {
        val egenAktivitetData = aktivitetData.egenAktivitetData
        return Egenaktivitet(
            aktivitetType = AktivitetType.egenaktivitet,
            avtaltMedNav = aktivitetData.isAvtalt(),
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            hensikt = egenAktivitetData.hensikt,
            oppfolging = egenAktivitetData.oppfolging
        )
    }

    private fun mapTilJobbsoeking(aktivitetData: AktivitetData): Jobbsoeking {
        val stillingsSoekAktivitetData = aktivitetData.stillingsSoekAktivitetData
        val stillingsoekEtikett = stillingsSoekAktivitetData.stillingsoekEtikett
        return Jobbsoeking(
            aktivitetType = AktivitetType.jobbsoeking,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            arbeidsgiver = stillingsSoekAktivitetData.arbeidsgiver,
            stillingsTittel = stillingsSoekAktivitetData.stillingsTittel,
            arbeidssted = stillingsSoekAktivitetData.arbeidssted,
            stillingsoekEtikett =
                stillingsoekEtikett
                    ?.let { it.name }
                    ?.let { Jobbsoeking.StillingsoekEtikett.valueOf(it) },
            kontaktPerson = stillingsSoekAktivitetData.kontaktPerson
        )
    }

    private fun mapTilSokeavtale(aktivitetData: AktivitetData): Sokeavtale {
        val sokeAvtaleAktivitetData = aktivitetData.sokeAvtaleAktivitetData
        return Sokeavtale(
            aktivitetType = AktivitetType.sokeavtale,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            antallStillingerSokes = sokeAvtaleAktivitetData.getAntallStillingerSokes(),
            antallStillingerIUken = sokeAvtaleAktivitetData.getAntallStillingerIUken(),
            avtaleOppfolging = sokeAvtaleAktivitetData.getAvtaleOppfolging(),
            )

    }

    private fun mapTilIjobb(aktivitetData: AktivitetData): Ijobb {
        val iJobbAktivitetData: IJobbAktivitetData? = aktivitetData.iJobbAktivitetData
        val jobbStatusType: JobbStatusTypeData? = iJobbAktivitetData?.jobbStatusType
        return Ijobb(
            aktivitetType = AktivitetType.ijobb,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            jobbStatusType = jobbStatusType?.let { it.name }?.let { Ijobb.JobbStatusType.valueOf(it) },
            ansettelsesforhold= iJobbAktivitetData?.ansettelsesforhold,
            arbeidstid= iJobbAktivitetData?.arbeidstid,
        )
    }

    private fun mapTilBehandling(aktivitetData: AktivitetData): Behandling {
        val behandlingAktivitetData = aktivitetData.behandlingAktivitetData
        return Behandling(
            aktivitetType = AktivitetType.behandling,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            behandlingType = behandlingAktivitetData.behandlingType,
            behandlingSted = behandlingAktivitetData.behandlingSted,
            effekt = behandlingAktivitetData.effekt,
            behandlingOppfolging = behandlingAktivitetData.behandlingOppfolging
        )
    }

    private fun mapTilMote(aktivitetData: AktivitetData): Mote {
        val moteData = aktivitetData.moteData
        return Mote(
            aktivitetType = AktivitetType.mote,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            adresse = moteData.adresse,
            forberedelser = moteData.forberedelser,
            kanal = Mote.Kanal.valueOf(moteData.kanal.name),
            referat = moteData.referat,
            referatPublisert = moteData.isReferatPublisert
        )
    }

    private fun mapTilSamtalereferat(aktivitetData: AktivitetData): Samtalereferat {
        val moteData = aktivitetData.moteData
        return Samtalereferat(
            aktivitetType = AktivitetType.samtalereferat,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            kanal = Samtalereferat.Kanal.valueOf(moteData.kanal.name),
            referat = moteData.referat,
            referatPublisert = moteData.isReferatPublisert
        )
    }

    private fun mapTilStillingFraNav(aktivitetData: AktivitetData): StillingFraNav {
        val stillingFraNavData = aktivitetData.stillingFraNavData
        val stillingFraNavCvKanDelesData: StillingFraNavAllOfCvKanDelesData? =
            mapCvKanDelesData(stillingFraNavData.cvKanDelesData)
        val soknadsstatusEnum = stillingFraNavData.soknadsstatus
            ?.let { it.name }?.let { StillingFraNav.Soknadsstatus.valueOf(it) }

        return StillingFraNav(
            aktivitetType = AktivitetType.stillingFraNav,
            avtaltMedNav = aktivitetData.isAvtalt,
            aktivitetId = aktivitetData.id.toString(),
            kontorsperreEnhetId = aktivitetData.kontorsperreEnhetId,
            oppfolgingsperiodeId = aktivitetData.oppfolgingsperiodeId,
            status = Status.valueOf(aktivitetData.status.name),
            beskrivelse = aktivitetData.beskrivelse,
            tittel = aktivitetData.tittel,
            fraDato = DateUtils.toOffsetDateTime(aktivitetData.fraDato),
            tilDato = DateUtils.toOffsetDateTime(aktivitetData.tilDato),
            opprettetDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),
            endretDato = DateUtils.toOffsetDateTime(aktivitetData.opprettetDato),

            cvKanDelesData = stillingFraNavCvKanDelesData,
            soknadsfrist = stillingFraNavData.soknadsfrist,
            svarfrist = DateUtils.toLocalDate(stillingFraNavData.svarfrist),
            arbeidsgiver = stillingFraNavData.arbeidsgiver,
            bestillingsId = stillingFraNavData.bestillingsId,
            stillingsId = stillingFraNavData.stillingsId,
            arbeidssted = stillingFraNavData.arbeidssted,
            soknadsstatus = soknadsstatusEnum
        )
    }

    private fun mapCvKanDelesData(cvKanDelesData: CvKanDelesData?): StillingFraNavAllOfCvKanDelesData? {
        if (cvKanDelesData == null) return null
        return StillingFraNavAllOfCvKanDelesData(
            kanDeles = cvKanDelesData.kanDeles,
            endretTidspunkt = DateUtils.toOffsetDateTime(cvKanDelesData.endretTidspunkt),
            endretAv = cvKanDelesData.endretAv,
            endretAvType = StillingFraNavAllOfCvKanDelesData.EndretAvType.valueOf(cvKanDelesData.endretAvType.name),
            avtaltDato = DateUtils.toLocalDate(cvKanDelesData.avtaltDato)
        )
    }
}
